Clean up ``AsyncHTTPProvider`` instantiation

fselmo opened the pull request:
### What was wrong?

Related to #1416. This is listed as non-breaking but instead of having an `async_net` module, it seems more appropriate to set the `net` module up as the `AsyncNet` class. This will break things.

### How was it fixed?

- Add ``is_async`` flag to base provider classes and appropriately set them to ``True`` or ``False``.
- Use these flags to determine whether to load sync or async default middlewares and default modules.
- Refactor the ``Web3`` property ``async_net`` as ``net`` mapped to ``AsyncNet`` class when dealing with async and ``Net`` when dealing with sync.

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)
- [x] Update docs

#### Cute Animal Picture

 ___ O>
/\ /\

fselmo committed:
Clean up ``AsyncHTTPProvider`` instantiation

- Add ``is_async`` flag to base provider classes and appropriately set them to ``True`` or ``False``.
- Use these flags to determine whether to load sync or async default middlewares and default modules.

fselmo committed:
Clean up ``async_net`` as a separate mypy type from ``net``

- Also minor cleanup that came along with these changes: e.g. expect a more specific error, ``NameNotFound`` for ens test since the async version started to wrongly xpass.

fselmo committed:
Update documentation for ``AsyncHTTPProvider`` instantiation.

pacrob reviewed (approved):
one small update noted, otherwise lgtm! thanks!

fselmo committed:
Clean up from comment on PR #2736

fselmo committed:
Some cleanup. Leave async typing decisions to upcoming PR.

- No need for the ``Net`` module to be the only one to have to `cast` values because the `net` property on `Web3` is a `Union[Net, AsyncNet]`. Instead, set it like the others and make one decision for all at the same time in an upcoming PR.
